src/main/java/org/interledger/ilp/exceptions/InterledgerException.java

```diff
@@ -1,36 +1,207 @@
+     * Check the RFC https://interledger.org/rfcs/0003-interledger-protocol/#errors
+     * for the newest updated doc.
+     * 
+     * @param errCode,    one of the defined InterledgerException.ErrorCode defined.
+     * @param triggeredBy ILP address (prefix if it's a ledger) of the entity that originally emitted the error
+     * @param forwardedBy ILP address list of connectors that relayed the error message
+     * @param triggeredAt Time when the error was initially emitted
+     * @param data        OCTET STRINGSIZE(0..8192) Unless otherwise specified, data SHOULD be encoded as UTF-8. 
+     *                    Protocols built on top of ILP SHOULD specify the encoding format of error data 
+     */
+    public InterledgerException(
+        ErrorCode errCode,
+        InterledgerAddress triggeredBy,
+        ArrayList<InterledgerAddress> forwardedBy,
+        ZonedDateTime triggeredAt, 
+        String data) {
+        super();
+        if (errCode == null || triggeredBy == null || forwardedBy == null || data == null) {
+            throw new IllegalArgumentException("parameters can NOT be null");
+        }
+        this.errCode     = errCode;
+        this.triggeredBy = triggeredBy;
+        this.forwardedBy = forwardedBy;
+        this.triggeredAt = triggeredAt;
+        this.data        = data;
+    }
+    
+    /**
+     * Constructs an instance of <code>InterledgerException</code>
+     * with default parameters for forwardedBy (Empty list) and 
+     * triggeredAt (ZonedDateTime.now()). 
+     *  In most situations such values match the default ones 
+     * when triggering a new exception (vs an exception received
+     * from another ILP node that is being forwarded back to 
+     * originating request clients)
+     * 
+     * Check the RFC https://interledger.org/rfcs/0003-interledger-protocol/#errors
+     * for the newest updated doc.
+     */
+    public InterledgerException(
+        ErrorCode errCode,
+        InterledgerAddress triggeredBy,
+        String data) {
+        this(errCode, triggeredBy, EMPTY_ForwardedBy_List, ZonedDateTime.now(), data);
+    }
+
+    public ErrorCode getErrCode() {
+        return errCode;
+    }
+
+    /**
+     * @return ErrorType categorizes the error as FINAL, TEMPORARY or RELATIVE
+     *         (See RFC for more info)
+     */
+    public ErrorType getErrorType(){
+        return errCode.errorType;
+    }
+
+    public InterledgerAddress getTriggeredBy() {
+        return triggeredBy;
+    }
+
+    public ArrayList<InterledgerAddress> getForwardedBy() {
+        return forwardedBy;
+    }
+
+    public ZonedDateTime getTriggeredAt() {
+        return triggeredAt;
+    }
+
+    public String getData() {
+        return data;
+    }
+
+    /**
+     * Used by connectors to add themself to the forwaredBy list
+     * @param selfAdd connector ILP address.
+     */
+    public void addSelfToForwardedByList(InterledgerAddress selfAdd) {
```

Comment:
Argg Threads!!! I always forget about threads! (I miss single-threaded scripting code now)
I just opted for the simple secure probe solution. Use an inmutable InterledgerAddress[] inmutable object.

I notice (correct me if I'm wrong) that there will be two scenarios:
- A node (ledger/connector/...) raise a new exception. In that case forwaredBy will be an empty array.
- A connector receives the exception, add itself  to the forwardedBy list and send back to other ILP nodes.  In this case I added an static utility method addSelfToForwardedBy to "easify" the use of the constructor.
 As a bonus this helper/utility static method check for "loops" (the connector find himself already added to the forwaredBy list) and throws a RuntimeException.

Not 100% sure what to do when detecting a loop but launching an exception is safer than ignoring it. This probably is a sympton of something really wrong in the routing setup.